src/main/java/org/folio/service/transactions/BatchTransactionService.java

```diff
@@ -54,25 +94,40 @@ private void checkIdIsPresent(List<Transaction> transactions, String operation)
       if (StringUtils.isBlank(transaction.getId())) {
         throw new HttpException(400, String.format("Id is required in transactions to %s.", operation));
       }
+      if (TYPES_WITH_SUMMARIES.contains(transaction.getTransactionType()) && transaction.getSourceInvoiceId() == null &&
+          (transaction.getEncumbrance() == null || transaction.getEncumbrance().getSourcePurchaseOrderId() == null)) {
+        throw new HttpException(400, String.format("Missing invoice id or order id in transaction %s", transaction.getId()));
+      }
     }
   }
 
-  private Future<Void> createTransactions(List<Transaction> transactions, DBConn conn) {
+  private Future<Void> createTransactions(List<Transaction> transactions, DBConn conn,
+      Map<String, String> okapiHeaders) {
     if (transactions.isEmpty())
       return Future.succeededFuture();
-    return Future.failedFuture(new HttpException(500, "transactionsToCreate: not implemented"));
+    populateMetadata(transactions, okapiHeaders);
+    return createOrUpdate(transactions,
+        (transactionService, transaction) -> transactionService.createTransaction(transaction, conn).mapEmpty(), conn)
+      .onSuccess(v -> logger.info("Batch transactions: successfully created transactions"))
+      .onFailure(t -> logger.error("Batch transactions: failed to create transactions", t));
   }
 
   private Future<Void> updateTransactions(List<Transaction> transactions, DBConn conn) {
     if (transactions.isEmpty())
       return Future.succeededFuture();
-    return Future.failedFuture(new HttpException(500, "transactionsToUpdate: not implemented"));
+    return createOrUpdate(transactions,
+        (transactionService, transaction) -> transactionService.updateTransaction(transaction, conn), conn)
+      .onSuccess(v -> logger.info("Batch transactions: successfully updated transactions"))
+      .onFailure(t -> logger.error("Batch transactions: failed to update transactions", t));
   }
 
   private Future<Void> deleteTransactions(List<String> ids, DBConn conn) {
-    if (ids.isEmpty())
-      return Future.succeededFuture();
-    return Future.failedFuture(new HttpException(500, "idsOfTransactionsToDelete: not implemented"));
+    Future<Void> f = Future.succeededFuture();
+    for (String id : ids) {
+      f = f.compose(v -> defaultTransactionService.deleteTransactionById(id, conn));
```

Comment:
I hope that it might be helpfull : https://github.com/folio-org/mod-finance-storage/blob/43139a0f11bbf00268009fab7942ca5aadbf1e17/src/main/java/org/folio/rest/persist/HelperUtils.java#L158
please try to use with chainCall static method. It will increase readability

Reply:
I tried to explain in the "Learning" section that the method used by chainCall was not reliable. But I guess I should fix it and use it here.